snuba/split.py

```diff
@@ -9,86 +9,155 @@
 # queries before hitting the 90d limit (2+20+200+2000 hours == 92 days).
 STEP_GROWTH = 10
 
+
 def split_query(query_func):
-    """
-    If a query is:
-        - ORDER BY timestamp DESC
-        - has no grouping
-        - has an offset/limit
-        - has a large time range
-    We know we have to reverse-sort the entire set of rows to return the small
-    chunk at the end of the time range, so optimistically split the time range
-    into smaller increments, and start with the last one, so that we can potentially
-    avoid querying the entire range.
-    """
     def wrapper(*args, **kwargs):
         body = args[0]
-        use_split, date_align, split_step = state.get_configs([
+        use_split = state.get_configs([
             ('use_split', 0),
-            ('date_align_seconds', 1),
-            ('split_step', 3600),  # default 1 hour
         ])
-        to_date = util.parse_datetime(body['to_date'], date_align)
-        from_date = util.parse_datetime(body['from_date'], date_align)
         limit = body.get('limit', 0)
         remaining_offset = body.get('offset', 0)
         orderby = util.to_list(body.get('orderby'))
 
+        total_col_count = len(util.all_referenced_columns(body))
+        min_col_query = dict(body, **{'selected_columns': ['project_id', 'event_id', 'timestamp']})
+        min_col_count = len(util.all_referenced_columns(min_col_query))
+
+        if (
+            use_split and limit
```

Comment:
There are a few conditions that are shared here and need to be `True` for us to even consider splitting. I think it'd be clearer if we checked those first (specifically `use_split`, `limit` and `body.get('groupby')` and early returned if *any* type of split can't/shouldn't be performed, or restructure this like:

```python
if common conditions:
  if column conditions:
    return col_split(…)
  elif time conditions:
    return time_split(…)

return query_func(…)
```

It's also cheaper to check the time conditions than the column conditions (since time conditions don't require building a synthetic query) so I'd consider switching the order of evaluation.

Reply:
Added the common conditions. Talked to JT about preferring col_split over time_split if both sets of conditions are met though, hence the ordering.

Reply:
Oh, right — good point. Might be worth a comment in the code about why the ordering is significant since it's not immediately obvious.